.github/workflows/main.yaml

```diff
@@ -61,38 +78,91 @@ jobs:
           aws-region: ${{ secrets.AWS_DEFAULT_REGION }}
       - name: Get cache
         run: aws s3 sync s3://docker-buildx-cache/emqx-builder/${{ matrix.platform[0] }} /tmp/.docker-buildx-cache
-      - uses: docker/setup-buildx-action@v2
-      - uses: docker/setup-qemu-action@v2
-      - uses: docker/login-action@v2
+      - uses: docker/setup-buildx-action@v3
+      - uses: docker/setup-qemu-action@v3
+      - uses: docker/login-action@v3
         with:
           registry: ghcr.io
           username: ${{ github.actor }}
           password: ${{ github.token }}
-      - uses: docker/metadata-action@v4
+      - uses: docker/metadata-action@v5
         id: base_meta
         with:
-          images: ghcr.io/${{ github.repository }}/base-${{ matrix.base_image_vsn }}
+          images: ${{ env.REGISTRY_IMAGE }}
           tags: type=raw,value=${{ matrix.platform[0] }}
       - name: Build base image
-        uses: docker/build-push-action@v3
+        id: build
+        uses: docker/build-push-action@v5
         with:
-          push: true
           pull: true
           cache-from: type=local,src=/tmp/.docker-buildx-cache,mode=max
           cache-to: type=local,dest=/tmp/.docker-buildx-cache-new,mode=max
           platforms: ${{ matrix.platform[1] }}
           tags: ${{ steps.base_meta.outputs.tags }}
           labels: ${{ steps.base_meta.outputs.labels }}
           file: ${{ matrix.platform[0] }}/Dockerfile
+          outputs: type=image,name=${{ env.REGISTRY_IMAGE }},push-by-digest=true,name-canonical=true,push=true
           context: .
       - name: Update cache
         run: aws s3 sync --delete /tmp/.docker-buildx-cache-new s3://docker-buildx-cache/emqx-builder/${{ matrix.platform[0] }}
+      - name: Export digests
+        run: |
+          mkdir -p /tmp/digests
+          digest="${{ steps.build.outputs.digest }}"
+          touch "/tmp/digests/${digest#sha256:}"
+      - name: Upload digest
+        uses: actions/upload-artifact@v3
+        with:
+          name: digests
+          path: /tmp/digests/*
+          if-no-files-found: error
+          retention-days: 1
 
-  build:
+  merge-base:
```

Comment:
Not sure if this is even needed. If the base image only used to produce the final one, perhaps we could just create arch-specific versions, include architecture in the image tag or even name, then generate corresponding `BUILD_FROM` argument in the build/build step below. What do you think?

Reply:
Makes sense.  It's only the final builder image that we might want to change the tagging scheme, but the intermediates might be refactored to have the arch in the tag so we avoid merging.  I'll adjust it.